comp/core/workloadmeta/collectors/internal/crio/containers.go

```diff
@@ -0,0 +1,298 @@
+}
+
+// getContainerStatus retrieves the status of a container.
+func getContainerStatus(ctx context.Context, client crio.Client, containerID string) (*v1.ContainerStatus, map[string]string) {
+	statusResponse, err := client.GetContainerStatus(ctx, containerID)
+	status := statusResponse.GetStatus()
+	info := statusResponse.GetInfo()
+	if err != nil || status == nil {
+		log.Errorf("Failed to get container status for container %s: %v", containerID, err)
+		return &v1.ContainerStatus{State: v1.ContainerState_CONTAINER_UNKNOWN}, make(map[string]string)
+	}
+	return status, info
+}
+
+// getResourceLimits extracts CPU and memory limits from container status or info as a fallback.
+func getResourceLimits(containerStatus *v1.ContainerStatus, info map[string]string) (*float64, *uint64) {
+	// First, try to get resources from containerStatus
+	if containerStatus != nil && containerStatus.GetResources() != nil && containerStatus.GetResources().GetLinux() != nil {
+		var cpuLimit *float64
+		var memLimit *uint64
+		cpuPeriod := float64(containerStatus.GetResources().GetLinux().GetCpuPeriod())
+		cpuQuota := float64(containerStatus.GetResources().GetLinux().GetCpuQuota())
+		memLimitInBytes := uint64(containerStatus.GetResources().GetLinux().GetMemoryLimitInBytes())
+
+		if cpuPeriod != 0 && cpuQuota != 0 {
+			limit := cpuQuota / cpuPeriod
+			cpuLimit = &limit
+		}
+		if memLimitInBytes != 0 {
+			memLimit = &memLimitInBytes
+		}
+		return cpuLimit, memLimit
+	}
+
+	if info == nil || info["info"] == "" {
+		log.Warn("Info map is nil or does not contain resource information")
+		return nil, nil
+	}
+
+	// Fallback to parsing resources from info if status resources are nil
+	var parsedInfo struct {
+		RuntimeSpec struct {
+			Linux struct {
+				Resources struct {
+					CPU struct {
+						Quota  int64 `json:"quota"`
+						Period int64 `json:"period"`
+					} `json:"cpu"`
+					Memory struct {
+						LimitInBytes int64 `json:"memoryLimitInBytes"`
+					} `json:"memory"`
+				} `json:"resources"`
+			} `json:"linux"`
+		} `json:"runtimeSpec"`
+	}
+
+	if err := json.Unmarshal([]byte(info["info"]), &parsedInfo); err != nil {
+		log.Warnf("Failed to parse resources from container info: %v", err)
+		return nil, nil
+	}
+
+	cpuPeriod := float64(parsedInfo.RuntimeSpec.Linux.Resources.CPU.Period)
+	cpuQuota := float64(parsedInfo.RuntimeSpec.Linux.Resources.CPU.Quota)
+	memLimitInBytes := uint64(parsedInfo.RuntimeSpec.Linux.Resources.Memory.LimitInBytes)
+
+	var cpuLimit *float64
+	var memLimit *uint64
+	if cpuPeriod != 0 && cpuQuota != 0 {
+		limit := cpuQuota / cpuPeriod
+		cpuLimit = &limit
+	}
+	if memLimitInBytes != 0 {
+		memLimit = &memLimitInBytes
+	}
+	return cpuLimit, memLimit
+}
+
+// getContainerImage retrieves and converts a container image to workloadmeta format.
+func getContainerImage(ctx context.Context, client crio.Client, imageSpec *v1.ImageSpec) workloadmeta.ContainerImage {
+	image, err := client.GetContainerImage(ctx, imageSpec)
```

Comment:
nit: minor optimization, but if `imageSpec` is `nil` then I would expect nothing would be returned, so we could probably short circuit this call in that case to prevent unnecessary queries to the crio socket